openwisp_monitoring/check/tests/test_iperf.py

```diff
@@ -0,0 +1,477 @@
+
+    def test_iperf_check_content_object_not_device(self):
+        check = Check(
+            name='Iperf check',
+            check_type=self._IPERF,
+            content_object=self._create_user(),
+            params={},
+        )
+        try:
+            check.check_instance.validate()
+        except ValidationError as e:
+            self.assertIn('device', str(e))
+        else:
+            self.fail('ValidationError not raised')
+
+    def test_iperf_check_schema_violation(self):
+        check, _ = self._create_iperf_test_env()
+        invalid_params = [
+            {'port': -1232},
+            {'time': 0},
+            {'port': 'invalid port'},
+            {'time': 'invalid time'},
+            {'port': '-12a'},
+            {'time': '3test22'},
+            {'port': 0},
+            {'port': 797979},
+            {'time': 36000},
+            {'port': ''},
+            {'time': ''},
+        ]
+        for invalid_params in invalid_params:
+            check.params = invalid_params
+            check.save()
+            try:
+                check.check_instance.validate()
+            except ValidationError as e:
+                self.assertIn('Invalid param', str(e))
+            else:
+                self.fail('ValidationError not raised')
+
+    def test_iperf_check(self):
+        check, _ = self._create_iperf_test_env()
+        error = "ash: iperf3: not found"
+        tcp_result = loads(RESULT_TCP)['end']
+        udp_result = loads(RESULT_UDP)['end']['sum']
+
+        with self.subTest('Test iperf3 is not installed on the device'):
+            with patch.object(
+                Iperf, '_exec_command'
+            ) as mock_exec_command, patch.object(
+                Iperf,
+                '_get_iperf_servers',
+                return_value=['iperf.openwisptestserver.com'],
+            ) as mock_get_iperf_servers:
+                mock_exec_command.side_effect = [(error, 127)]
+                with patch.object(iperf_logger, 'warning') as mock_warn:
+                    check.perform_check(store=False)
+                    mock_warn.assert_called_with(
+                        f'Iperf3 is not installed on the "{self.device}", error - {error}'
+                    )
+                self.assertEqual(mock_warn.call_count, 1)
+                self.assertEqual(mock_exec_command.call_count, 1)
+                mock_get_iperf_servers.assert_called_once_with(
+                    self.device.organization.id
+                )
+
+        with self.subTest('Test iperf check passes in both TCP & UDP'):
+            with patch.object(
+                Iperf, '_exec_command'
+            ) as mock_exec_command, patch.object(
+                Iperf,
+                '_get_iperf_servers',
+                return_value=['iperf.openwisptestserver.com'],
+            ) as mock_get_iperf_servers, patch.object(
+                iperf_logger, 'warning'
+            ) as mock_warn:
```

Comment:
I did this because if I use decorator style mocks for `test_iperf_check` (which consists of various subTests) then for every subTests I'm getting additive `mock_method` call counts (It includes method calls from previous subTest) which I thought is not a good idea.
**For eg:**
```py
    @patch.object(Iperf, '_exec_command')
    @patch.object(Iperf, '_get_iperf_servers', return_value=['iperf.openwisptestserver.com'])
    @patch.object(iperf_logger, 'warning')
    def test_iperf_check(self, mock_warn, mock_get_iperf_servers, mock_exec_command):
        check, _ = self._create_iperf_test_env()
        error = "ash: iperf3: not found"
        tcp_result = loads(RESULT_TCP)['end']
        udp_result = loads(RESULT_UDP)['end']['sum']

        with self.subTest('Test iperf3 is not installed on the device'):
                mock_exec_command.side_effect = [(error, 127)]
                with patch.object(iperf_logger, 'warning') as mock_warn:
                    check.perform_check(store=False)
                    mock_warn.assert_called_with(
                        f'Iperf3 is not installed on the "{self.device}", error - {error}'
                    )
                self.assertEqual(mock_warn.call_count, 1)
                self.assertEqual(mock_exec_command.call_count, 1)
                mock_get_iperf_servers.assert_called_once_with(
                    self.device.organization.id
                )
        with self.subTest('Test iperf check passes in both TCP & UDP'):
                mock_exec_command.side_effect = [(RESULT_TCP, 0), (RESULT_UDP, 0)]
                ...
                ...

                self.assertEqual(mock_warn.call_count, 0)
                # AssertionError: 1 != 0 (Includes call count from previous subTest)
                self.assertEqual(mock_exec_command.call_count, 2)
                # AssertionError: 3 != 2 (Same here)
    
```